### ISSUE TITLE
Better auth flow for react-router

### ISSUE THREAD

**Issue by choonkending** [choonkending/react-webpack-node#107]
# EDITED

## Goal of this issue fix
We want a good example of an authentication flow for universal redux and react-router. This will include:
- [x] Redirecting users to the Login page when they access a route/page which requires authentication.
    - [x] Server-side authentication redirects when the user accesses a route directly, e.g. `/dashboard` 
    - [x] Client-side redirects

# Scenarios

## Client side redirect
* Given I am not logged in
* And I am in the `/` route
* When I navigate to `/dashboard` which requires authentication
* Then I will be redirected to the `/login` path

## Server side redirect
* Given I am not logged in
* When I navigate to `/dashboard` which requires authentication
* Then I will be served the `/login` page

Possible example implementations:
https://gist.github.com/iNikNik/1eabfc6bcc132384368c

**Comment by choonkending** [comment]
https://github.com/rackt/react-router/issues/2387

**Comment by choonkending** [comment]
http://stackoverflow.com/questions/33358249/server-side-auth-flow-with-react-router?rq=1

**Comment by hoodsy** [comment]
@choonkending have you done any work on this? I'm looking to implement this in a personal project, and would love to help with / take a look at your implementation.

**Comment by choonkending** [comment]
@hoodsy I've been looking at different solutions. 

The main problem is dealing with server-side redirects (when a user is not authenticated) in a nice and understandable way. 

Really happy to get help/suggestions with tightening up the auth flow! Thanks a lot! So far https://gist.github.com/iNikNik/1eabfc6bcc132384368c looks to be a good one. 

I feel like I'm missing something simple/there could be a simple solution to this.

**Comment by hoodsy** [comment]
@choonkending I've been doing a bit of research as well, and liked the ```onEnter``` approach shown in that gist best. 

What do you think about only redirecting with redux-simple-router, for consistency sake? Using  ```pushPath``` all app routing could happen from the front end. 

I was also thinking that  ```onEnter``` would fire off an action to test  ```user.token```. The token, received from the server, could be persisted via localStorage (or redis) to maintain sessions. 

So the flow would be something like this:
Login -> auth action -> server responds with token -> token stored in user -> onEnter action called for route change -> if user.token, gain access -> request to server for route data (carrying token) -> server response, if token is valid, receive route data 

I hope this is clear, and it would be great to hear your thoughts!

**Comment by choonkending** [comment]
@hoodsy Sorry for the late reply.

> Login -> auth action -> server responds with token -> token stored in user -> onEnter action called for route change -> if user.token, gain access -> request to server for route data (carrying token) -> server response, if token is valid, receive route data

Sounds like a great idea! I did not consider doing it that way (server checking for token and redirecting). That greatly simplifies things imo. 

>  authentication only on the server with Passport's cookies.

I'm not sure if pushPath works on the server? I tried it a while ago and had trouble with it (maybe I just did it wrong). I'll have a look at it tonight.

This is exciting! I really REALLY wanna get this fixed.

**Comment by hoodsy** [comment]
@choonkending I'll work on that example.

As I have it now, a server redirect will send them to ```/login``` if any request isn't authenticated (similar to what you have).

```javascript
export function isAuthenticated(req, res, next) {
  if (req.isAuthenticated()) next()
  else res.status(401).redirect('/login')
}

// Routes requiring authentication
app.all('/api/*', isAuthenticated)
app.get('/', isAuthenticated, initialRender)
```

So on the initial load, they get to the login page if not authenticated. From login, I'm not sure how to get around using ```pushPath('/')``` from my LOGIN_USER_SUCCESS action to redirect to the main view.

The redux middleware is to handle an expired session - in case it is invalid, they are redirected to the login. But I guess that is now handled on the server anyways... The middleware seems redundant.

**Comment by choonkending** [comment]
@hoodsy Thanks for explaining this! 

>  From login, I'm not sure how to get around using pushPath('/') from my LOGIN_USER_SUCCESS action to redirect to the main view.

Could you dispatch an pushPath, similar to 
```javascript
dispatch(pushPath(`/login?next=${this.props.location.pathname}`));
```

It's a bit tricky isn't it? I'll have a long think about what the right way to do it is today. (But I'm also glad you're helping out, this one has been killing me).

**Comment by hoodsy** [comment]
@choonkending look forward to hearing your thoughts! You can check out my
approach at:
https://github.com/hoodsy/anchor

**Comment by choonkending** [comment]
@hoodsy For your approach, do you plan on restricting all access to your routes?

``` javascript
    router.all('/api/*', isAuthenticated)
    router.get('/', isAuthenticated, initialRender)
```

**Comment by hoodsy** [comment]
@choonkending
To me, the route restriction was somewhat counter intuitive. I am
restricted to /login, and get redirected to it, until logging in or
registering.

From the routes it seems as though /login would be restricted as well, but
it didn't seem to be the case when I was testing it. I am thinking this is
due to react router.

**Comment by choonkending** [comment]
@hoodsy I have a semi-working solution now. Check the PR.

The solution I did was to use react-router's onEnter method for components. I supplied a requireAuth function [here](https://github.com/choonkending/react-webpack-node/pull/125/files#diff-c1858c355a655a4c2bbc1b0df3156942R23). It redirects successfully on unauthenticated requests.

Issue now is that `fetch` from the server side always returns 
```javascript
{authenticated: false}
``` 
So if you go to access the `/dashboard` url straight away, even if you are logged in, fetch returns the above. Maybe I'm missing something crucial. Like cookies or something. If we could somehow pass down react context, or store values, or even parameter values down from `match` to `onEnter`, this could be solved.

Interested to know if you have any opinions. Also @psimyn if you have any thoughts :hamburger:
